File: client/java/src/com/tibco/tgdb/query/impl/GremlinResult.java
package com.tibco.tgdb.query.impl;
/**
 * Copyright 2016 TIBCO Software Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except
 * in compliance with the License.
 * A copy of the License is included in the distribution package with this file.
 * You also may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * File name : TGQueryOption.java
 * SVN Id: $Id$
 */

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import com.tibco.tgdb.exception.TGException;
import com.tibco.tgdb.log.TGLogManager;
import com.tibco.tgdb.log.TGLogger;
import com.tibco.tgdb.model.TGAttribute;
import com.tibco.tgdb.model.TGEdge;
import com.tibco.tgdb.model.TGEntity;
import com.tibco.tgdb.model.TGNode;
import com.tibco.tgdb.model.impl.GraphObjectFactoryImpl;
import com.tibco.tgdb.model.impl.attribute.AbstractAttribute;
import com.tibco.tgdb.pdu.TGInputStream;

public class GremlinResult {

   	private enum ElementType {
       	Invalid(0),
       	List(1),
       	Attr(2),
       	AttrValue(3),
       	Entity(4),
       	Map(5);

    	private int id;
    
    	ElementType (int id) {
    		this.id = (short) id;
    	}

    	public static ElementType fromValue(int id) {
    		for (ElementType et : ElementType.values()) {
    			if (id == et.id) return et;
    		}
    		return Invalid;
    	}

    	public int getId() { 
    		return id; 
    	}
   	}

    static TGLogger gLogger        = TGLogManager.getInstance().getLogger();

    //Only handle entity right now
    static public void fillCollection(TGInputStream entityStream, GraphObjectFactoryImpl gof, Collection col) {
    	try {
    		//These types will have an enum to represent them
    		ElementType retType = ElementType.fromValue(entityStream.readByte());
    		if (retType == ElementType.List) {
    			constructList(entityStream, gof, col);
    		} else {
    			gLogger.log(TGLogger.TGLevel.Error, "Invalid gremlin response collection type : %d", retType.getId());
    		}
    	} catch (IOException ioe) {
            gLogger.log(TGLogger.TGLevel.Error, "Failed to read gremlin query response stream");
    	} catch (TGException tge) {
            gLogger.log(TGLogger.TGLevel.Error, "Failed to read gremlin query response stream");
    	}
    }
    
    static public void constructList(TGInputStream entityStream, GraphObjectFactoryImpl gof, Collection col) throws IOException, TGException {
   		int size = entityStream.readInt();
   		int et = entityStream.readByte();
   		ElementType elemType = ElementType.fromValue(et);
   		TGNode dummyNode = null;
   		if (elemType == ElementType.Attr || elemType == ElementType.AttrValue) {
   			dummyNode = gof.createNode();
   		}
   		for (int i=0; i<size; i++) {
   			if (elemType == ElementType.Entity) {
   				TGEntity.TGEntityKind kind = TGEntity.TGEntityKind.fromValue(entityStream.readByte());
   				if (kind == TGEntity.TGEntityKind.InvalidKind) {
   					gLogger.log(TGLogger.TGLevel.Error, "Invalid entity kind from gremlin response stream");
   					break;
   				}
   				if (kind == TGEntity.TGEntityKind.Node) {
   					TGNode node = gof.createNode();
   					node.readExternal(entityStream);
   					col.add(node);
   				}
   			} else if (elemType == ElementType.Attr) {
   				TGAttribute attr = AbstractAttribute.readExternal(dummyNode, entityStream);
   				col.add(attr);
   			} else if (elemType == ElementType.AttrValue) {
   				TGAttribute attr = AbstractAttribute.readExternal(dummyNode, entityStream);
   				col.add(attr.getValue());
   			} else if (elemType == ElementType.List) {
   				Collection colElem = new ArrayList();
   				constructList(entityStream, gof, colElem);
   				col.add(colElem);
   			} else if (elemType == ElementType.Map) {
   				Map<String, Object> mapElem = new HashMap<String, Object>();
   				constructMap(entityStream, gof, mapElem);
   				col.add(mapElem);
   			} else {
				gLogger.log(TGLogger.TGLevel.Error, "Invalid element type %d from gremlin response stream", et);
   			}
   		}
    }

    static public void constructMap(TGInputStream entityStream, GraphObjectFactoryImpl gof, Map<String, Object> map) throws IOException, TGException {
   		int size = entityStream.readInt();
   		for (int i=0; i<size; i++) {
   			String key = entityStream.readUTF();
			int et = entityStream.readByte();
			ElementType elemType = ElementType.fromValue(et);
			TGNode dummyNode = null;
			if (elemType == ElementType.Attr || elemType == ElementType.AttrValue) {
				dummyNode = gof.createNode();
			}
   			if (elemType == ElementType.Entity) {
   				TGEntity.TGEntityKind kind = TGEntity.TGEntityKind.fromValue(entityStream.readByte());
   				if (kind == TGEntity.TGEntityKind.InvalidKind) {
   					gLogger.log(TGLogger.TGLevel.Error, "Invalid entity kind from gremlin response stream");
   					break;
   				}
   				if (kind == TGEntity.TGEntityKind.Node) {
   					TGNode node = gof.createNode();
   					node.readExternal(entityStream);
   					map.put(key, node);
   				} else if (kind == TGEntity.TGEntityKind.Node) {
   					//It has no from/or node right now
   					TGEntity edge = gof.createEntity(TGEntity.TGEntityKind.Edge);
   					edge.readExternal(entityStream);
   					map.put(key, edge);
   				}
   			} else if (elemType == ElementType.Attr) {
   				TGAttribute attr = AbstractAttribute.readExternal(dummyNode, entityStream);
   				map.put(key, attr);
   			} else if (elemType == ElementType.AttrValue) {
   				TGAttribute attr = AbstractAttribute.readExternal(dummyNode, entityStream);
   				map.put(key, attr.getValue());
   			} else if (elemType == ElementType.List) {
   				Collection colElem = new ArrayList();
   				constructList(entityStream, gof, colElem);
   				map.put(key, colElem);
   			} else if (elemType == ElementType.Map) {
   				Map<String, Object> mapElem = new HashMap<String, Object>();
   				constructMap(entityStream, gof, mapElem);
   				map.put(key, mapElem);
   			} else {
				gLogger.log(TGLogger.TGLevel.Error, "Invalid element type %d from gremlin response stream", et);
   			}
   		}
    }
}

File: client/goAPI/query/GremlinResult.go
/**
 * Copyright 2018-19 TIBCO Software Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except
 * in compliance with the License.
 * A copy of the License is included in the distribution package with this file.
 * You also may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF DirectionAny KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * File name: GremlinResult.go
 * Created on: Apr 27, 2019
 * Created by: achavan
 * SVN id: $id: $
 *
 */

package query

import (
	"bytes"
	"fmt"
	"github.com/TIBCOSoftware/tgdb-client/client/goAPI/exception"
	"github.com/TIBCOSoftware/tgdb-client/client/goAPI/iostream"
	"github.com/TIBCOSoftware/tgdb-client/client/goAPI/model"
	"github.com/TIBCOSoftware/tgdb-client/client/goAPI/types"
)

// ======= Various Element Types for Gremlin Results =======
type ElementType int

const (
	ElementTypeInvalid ElementType = iota
	ElementTypeList
	ElementTypeAttr
	ElementTypeAttrValue
	ElementTypeAttrValueTransient
	ElementTypeEntity
	ElementTypeMap
)

func (elementType ElementType) String() string {
	// Use a buffer for efficient string concatenation
	var buffer bytes.Buffer

	if elementType&ElementTypeInvalid == ElementTypeInvalid {
		buffer.WriteString("ElementTypeInvalid")
	} else if elementType&ElementTypeList == ElementTypeList {
		buffer.WriteString("ElementTypeList")
	} else if elementType&ElementTypeAttr == ElementTypeAttr {
		buffer.WriteString("ElementTypeAttr")
	} else if elementType&ElementTypeAttrValue == ElementTypeAttrValue {
		buffer.WriteString("ElementTypeAttrValue")
	} else if elementType&ElementTypeAttrValueTransient == ElementTypeAttrValueTransient {
		buffer.WriteString("ElementTypeAttrValueTransient")
	} else if elementType&ElementTypeEntity == ElementTypeEntity {
		buffer.WriteString("ElementTypeEntity")
	} else if elementType&ElementTypeMap == ElementTypeMap {
		buffer.WriteString("ElementTypeMap")
	}
	if buffer.Len() == 0 {
		return ""
	}
	return buffer.String()
}

/////////////////////////////////////////////////////////////////
// Helper functions for Gremlin Result
/////////////////////////////////////////////////////////////////

func FillCollection(entityStream types.TGInputStream, gof types.TGGraphObjectFactory, col []interface{}) types.TGError {
	//logger.Log(fmt.Sprint("Entering GremlinResult:FillCollection"))
	eleType, err := entityStream.(*iostream.ProtocolDataInputStream).ReadByte()
	if err != nil {
		logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:FillCollection - unable to read eleType in the response stream w/ error: '%s'", err.Error()))
		errMsg := "GremlinResult:FillCollection - unable to read element type in the response stream"
		return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
	}
	logger.Debug(fmt.Sprintf("Inside Returning GremlinResult:FillCollection extracted eleType: '%+v'", eleType))

	if ElementType(eleType) == ElementTypeList {
		return ConstructList(entityStream, gof, col)
	} else {
		logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:FillCollection - Invalid gremlin response collection type : %+v", ElementType(eleType)))
		errMsg := fmt.Sprintf("GremlinResult:FillCollection - Invalid gremlin response collection type : %+v", ElementType(eleType))
		return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, "")
	}
	return nil
}

func ConstructList(entityStream types.TGInputStream, gof types.TGGraphObjectFactory, col []interface{}) types.TGError {
	logger.Log(fmt.Sprint("Entering GremlinResult:ConstructList"))
	size, err := entityStream.(*iostream.ProtocolDataInputStream).ReadInt()
	if err != nil {
		logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructList - unable to read size in the response stream w/ error: '%s'", err.Error()))
		errMsg := "GremlinResult:ConstructList unable to read size in the response stream"
		return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
	}
	logger.Debug(fmt.Sprintf("Inside Returning GremlinResult:FillCollection extracted size: '%+v'", size))

	eleType, err := entityStream.(*iostream.ProtocolDataInputStream).ReadByte()
	if err != nil {
		logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructList - unable to read eleType in the response stream w/ error: '%s'", err.Error()))
		errMsg := "GremlinResult:ConstructList unable to read element type in the response stream"
		return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
	}
	logger.Debug(fmt.Sprintf("Inside Returning GremlinResult:ConstructList extracted eleType: '%+v'", eleType))

	var dummyNode types.TGNode
	if ElementType(eleType) == ElementTypeAttr || ElementType(eleType) == ElementTypeAttrValue || ElementType(eleType) == ElementTypeAttrValueTransient {
		node1, err := gof.CreateNode()
		if err != nil {
			logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructList - unable to gof.CreateNode() for element type: '%+v'", ElementType(eleType)))
			errMsg := fmt.Sprintf("GremlinResult:ConstructList unable to create node for element type: '%+v'", ElementType(eleType))
			return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
		}
		dummyNode = node1
	}

	for i:=0; i<size; i++ {
		if ElementType(eleType) == ElementTypeEntity {
			entityType, err := entityStream.(*iostream.ProtocolDataInputStream).ReadByte()
			if err != nil {
				logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructList  - unable to read entityType in the response stream w/ error: '%s'", err.Error()))
				errMsg := "GremlinResult:ConstructList - unable to read entity type in the entity stream"
				return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
			}
			kindId := types.TGEntityKind(entityType)
			switch kindId {
			case types.EntityKindNode:
				node, nErr := gof.CreateNode()
				if nErr != nil {
					logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructList - unable to gof.CreateNode() for element type: '%+v'", ElementType(eleType)))
					errMsg := fmt.Sprintf("GremlinResult:ConstructList unable to create node for element type: '%+v'", ElementType(eleType))
					return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, nErr.Error())
				}
				err = node.ReadExternal(entityStream)
				if err != nil {
					logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructList  - unable to node.ReadExternal w/ error: '%s'", err.Error()))
					errMsg := "GremlinResult:ConstructList - unable to node.ReadExternal in the entity stream"
					return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
				}
				col = append(col, node)
			case types.EntityKindEdge:
				edge, nErr := gof.CreateEntity(types.EntityKindEdge)
				if nErr != nil {
					logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructList - unable to gof.CreateNode() for element type: '%+v'", ElementType(eleType)))
					errMsg := fmt.Sprintf("GremlinResult:ConstructList unable to create node for element type: '%+v'", ElementType(eleType))
					return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, nErr.Error())
				}
				err = edge.ReadExternal(entityStream)
				if err != nil {
					logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructList  - unable to edge.ReadExternal w/ error: '%s'", err.Error()))
					errMsg := "GremlinResult:ConstructList - unable to edge.ReadExternal in the entity stream"
					return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
				}
				col = append(col, edge)
			case types.EntityKindGraph:
				fallthrough
			case types.EntityKindInvalid:
				fallthrough
			default:
				logger.Error(fmt.Sprint("ERROR: Returning GremlinResult:ConstructList - Invalid entity kind from gremlin response stream"))
				break
			}
		} else if ElementType(eleType) == ElementTypeList {
			colElem := make([]interface{}, 0)
			_ = ConstructList(entityStream, gof, colElem)
			col = append(col, colElem)
		} else if ElementType(eleType) == ElementTypeMap {
			mapElem := make(map[string]interface{}, 0)
			_ = ConstructMap(entityStream, gof, mapElem)
			col = append(col, mapElem)
		} else if ElementType(eleType) == ElementTypeAttr || ElementType(eleType) == ElementTypeAttrValue || ElementType(eleType) == ElementTypeAttrValueTransient {
			attr, err := model.ReadExternalForEntity(dummyNode, entityStream)
			if err != nil {
				logger.Error(fmt.Sprintf("ERROR: Returning AbstractEntity:AbstractEntityReadExternal - unable to read attr w/ Error: '%+v'", err.Error()))
				return err
			}
			if ElementType(eleType) == ElementTypeAttr {
				col = append(col, attr)
			} else {
				col = append(col, attr.GetValue())
			}
		} else {
			logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructList - Invalid element type '%+v' from Gremlin response stream", eleType))
		}
	}	// End of for loop

	logger.Log(fmt.Sprint("Returning GremlinResult:ConstructList"))
	return nil
}

func ConstructMap(entityStream types.TGInputStream, gof types.TGGraphObjectFactory, colMap map[string]interface{}) types.TGError {
	logger.Log(fmt.Sprint("Entering GremlinResult:ConstructMap"))
	size, err := entityStream.(*iostream.ProtocolDataInputStream).ReadInt()
	if err != nil {
		logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructMap - unable to read size in the response stream w/ error: '%s'", err.Error()))
		errMsg := "GremlinResult:ConstructMap unable to read size in the response stream"
		return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
	}
	logger.Debug(fmt.Sprintf("Inside Returning GremlinResult:FillCollection extracted size: '%+v'", size))

	for i:=0; i<size; i++ {
		key, err := entityStream.(*iostream.ProtocolDataInputStream).ReadUTF()
		if err != nil {
			logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructMap  - unable to read key in the response stream w/ error: '%s'", err.Error()))
			errMsg := "GremlinResult:ConstructMap - unable to read key in the entity stream"
			return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
		}
		logger.Debug(fmt.Sprintf("Inside Returning GremlinResult:FillCollection extracted key: '%+v'", key))

		eleType, err := entityStream.(*iostream.ProtocolDataInputStream).ReadByte()
		if err != nil {
			logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructMap - unable to read eleType in the response stream w/ error: '%s'", err.Error()))
			errMsg := "GremlinResult:ConstructMap unable to read element type in the response stream"
			return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
		}
		logger.Debug(fmt.Sprintf("Inside Returning GremlinResult:ConstructMap extracted eleType: '%+v'", eleType))

		var dummyNode types.TGNode
		if ElementType(eleType) == ElementTypeAttr || ElementType(eleType) == ElementTypeAttrValue || ElementType(eleType) == ElementTypeAttrValueTransient {
			node1, err := gof.CreateNode()
			if err != nil {
				logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructList - unable to gof.CreateNode() for element type: '%+v'", ElementType(eleType)))
				errMsg := fmt.Sprintf("GremlinResult:ConstructList unable to create node for element type: '%+v'", ElementType(eleType))
				return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
			}
			dummyNode = node1
		}

		if ElementType(eleType) == ElementTypeEntity {
			entityType, err := entityStream.(*iostream.ProtocolDataInputStream).ReadByte()
			if err != nil {
				logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructMap  - unable to read entityType in the response stream w/ error: '%s'", err.Error()))
				errMsg := "GremlinResult:ConstructMap - unable to read entity type in the entity stream"
				return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
			}
			kindId := types.TGEntityKind(entityType)

			switch kindId {
			case types.EntityKindNode:
				node, nErr := gof.CreateNode()
				if nErr != nil {
					logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructMap - unable to gof.CreateNode() for element type: '%+v'", ElementType(eleType)))
					errMsg := fmt.Sprintf("GremlinResult:ConstructMap unable to create node for element type: '%+v'", ElementType(eleType))
					return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, nErr.Error())
				}
				err = node.ReadExternal(entityStream)
				if err != nil {
					logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructMap  - unable to node.ReadExternal w/ error: '%s'", err.Error()))
					errMsg := "GremlinResult:ConstructMap - unable to node.ReadExternal in the entity stream"
					return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
				}
				colMap[key] = node
			case types.EntityKindEdge:
				edge, nErr := gof.CreateEntity(types.EntityKindEdge)
				if nErr != nil {
					logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructMap - unable to gof.CreateNode() for element type: '%+v'", ElementType(eleType)))
					errMsg := fmt.Sprintf("GremlinResult:ConstructMap unable to create node for element type: '%+v'", ElementType(eleType))
					return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, nErr.Error())
				}
				err = edge.ReadExternal(entityStream)
				if err != nil {
					logger.Error(fmt.Sprintf("ERROR: Returning GremlinResult:ConstructMap  - unable to edge.ReadExternal w/ error: '%s'", err.Error()))
					errMsg := "GremlinResult:ConstructMap - unable to edge.ReadExternal in the entity stream"
					return exception.GetErrorByType(types.TGErrorGeneralException, "", errMsg, err.Error())
				}
				colMap[key] = edge
			case types.EntityKindGraph:
				fallthrough
			case types.EntityKindInvalid:
				fallthrough
			default:
				logger.Error(fmt.Sprint("ERROR: Returning GremlinResult:ConstructList - Invalid entity kind from gremlin response stream"))
				break
			}
		} else if ElementType(eleType) == ElementTypeAttr || ElementType(eleType) == ElementTypeAttrValue || ElementType(eleType) == ElementTypeAttrValueTransient {
			attr, err := model.ReadExternalForEntity(dummyNode, entityStream)
			if err != nil {
				logger.Error(fmt.Sprintf("ERROR: Returning AbstractEntity:AbstractEntityReadExternal - unable to read attr w/ Error: '%+v'", err.Error()))
				return err
			}
			colMap[key] = attr
			if ElementType(eleType) == ElementTypeAttr {
				colMap[key] = attr
			} else {
				colMap[key] = attr.GetValue()
			}
		} else if ElementType(eleType) == ElementTypeList {
			colElem := make([]interface{}, 0)
			_ = ConstructList(entityStream, gof, colElem)
			colMap[key] = colElem
		} else if ElementType(eleType) == ElementTypeMap {
			mapElem := make(map[string]interface{}, 0)
			_ = ConstructMap(entityStream, gof, mapElem)
			colMap[key] = mapElem
		}
	}	// End of for loop

	logger.Log(fmt.Sprint("Returning GremlinResult:ConstructMap"))
	return nil
}

